File: 프로그래머스/1/150370. 개인정보 수집 유효기간/개인정보 수집 유효기간.java
import java.util.*;

class Solution {
    private String plusMonth(String targetDate, int monthsNum) {
        String[] parts = targetDate.split("\\.");
        int year = Integer.parseInt(parts[0]);
        int month = Integer.parseInt(parts[1]);
        int day = Integer.parseInt(parts[2]);
        
        month += monthsNum;
        year += (month - 1) / 12;
        month = (month - 1) % 12 + 1;
        
        return String.format("%04d.%02d.%02d", year, month, day);
    }
    
    private boolean checkExpired(String today, String expiredDate) {
        return toCamparableDate(today) >= toCamparableDate(expiredDate);
    }
    
    private int toCamparableDate(String dateStr) {
        return Integer.parseInt(dateStr.replace(".", ""));
    }
    
    public int[] solution(String today, String[] terms, String[] privacies) {
        Map<String, Integer> termMap = new HashMap<>();
        for (String term : terms) {
            String[] parts = term.split(" ");
            termMap.put(parts[0], Integer.parseInt(parts[1]));
        }
        
        List<Integer> targets = new ArrayList<>();
        for (int i = 0; i < privacies.length; i++) {
            String[] parts = privacies[i].split(" ");
            String collectedDate = parts[0];
            String expiredDate = plusMonth(parts[0], termMap.get(parts[1]));
            
            if (checkExpired(today, expiredDate)) {
                targets.add(i + 1);
            }
        }
        
        return targets.stream()
            .mapToInt(Integer::intValue)
            .toArray();
    }
}

File: 프로그래머스/1/150370. 개인정보 수집 유효기간/개인정보 수집 유효기간.kt
class Solution {
    private fun plusMonth(targetDate: String, monthsNum: Int): String {
        val (yearStr, monthStr, dayStr) = targetDate.split(".")
        var year = yearStr.toInt()
        var month = monthStr.toInt()
        var day = dayStr.toInt()
        
        month += monthsNum
        year += (month - 1) / 12
        month = (month - 1) % 12 + 1
        
        return "%04d.%02d.%02d".format(year, month, day)
    }
    
    private fun checkExpired(today: String, expiredDate: String): Boolean {
        return toComparableDate(today) >= toComparableDate(expiredDate)
    }
    
    private fun toComparableDate(dateStr: String): Int {
        return dateStr.replace(".", "").toInt()
    }
    
    fun solution(today: String, terms: Array<String>, privacies: Array<String>): IntArray {
        val termMap = mutableMapOf<String, Int>()
        for (term in terms) {
            val (key, value) = term.split(" ")
            termMap[key] = value.toInt()
        }
        
        val targets = mutableListOf<Int>()
        for ((idx, privacy) in privacies.withIndex()) {
            val (collectedDate, termType) = privacy.split(" ")
            val expiredDate = plusMonth(collectedDate, termMap[termType]!!)
            
            if (checkExpired(today, expiredDate)) {
                targets.add(idx + 1)
            }
        }
        
        return targets.toIntArray()
    }
}